oprf/oprf.go

```diff
@@ -0,0 +1,267 @@
+package oprf
+
+import (
+	"crypto/sha512"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/cloudflare/circl/oprf/group"
+)
+
+const (
+	// OPRFCurve25519 is the constant to represent the OPRF curve25519 with SHA-512 (ELL2-RO) group
+	OPRFCurve25519 uint16 = 0x0001
+	// OPRFCurve448 is the constant to represent the OPRF curve448 with SHA-512 (ELL2-RO) group
+	OPRFCurve448 uint16 = 0x0002
+	// OPRFP256 is the constant to represent the OPRF P-256 with SHA-512 (SSWU-RO) group
+	OPRFP256 uint16 = 0x0003
+	// OPRFP384 is the constant to represent the OPRF P-384 with SHA-512 (SSWU-RO) group
+	OPRFP384 uint16 = 0x0004
+	// OPRFP521 is the constant to represent the OPRF P-521 with SHA-512 (SSWU-RO) group
+	OPRFP521 uint16 = 0x0005
+)
+
+var (
+	// ErrUnsupportedGroup is
+	ErrUnsupportedGroup = errors.New("The chosen group is not supported")
+)
+
+// BlindToken corresponds to a token that has been blinded.
+// Internally, it is a serialized Element.
+type BlindToken []byte
+
+// IssuedToken corresponds to a token that has been issued.
+// Internally, it is a serialized Element.
+type IssuedToken []byte
+
+// Token is the object issuance of the protocol.
+type Token struct {
+	data  []byte
+	blind *group.Scalar
+}
+
+// Evaluation corresponds to the evaluation over a token.
+type Evaluation struct {
+	element []byte
+}
+
+// PublicKey is a struct representing a public key.
+type PublicKey *group.Point
+
+// PrivateKey is a struct representing a private key.
+type PrivateKey *group.Scalar
+
+// KeyPair is an struct containing a public and private key.
+type KeyPair struct {
+	PubK  PublicKey
+	PrivK PrivateKey
+}
+
+// Client is a representation of a Client during protocol execution.
+type Client struct {
+	suite *group.Ciphersuite
+	ctx   string
+}
+
+// ClientContext implements the functionality of a Client.
+type ClientContext interface {
+	// Blind generates a token and blinded data.
+	Blind(in []byte) (*Token, *BlindToken, error)
+
+	// Unblind unblinds the server response.
+	Unblind(t *Token, e *Evaluation) (IssuedToken, error)
+
+	// Finalize outputs a byte array that corresponds to its input.
+	Finalize(t *Token, issuedT IssuedToken, info []byte) []byte
+}
+
+// Server is a representation of a Server during protocol execution.
+type Server struct {
+	suite *group.Ciphersuite
+	ctx   string
+	K     *KeyPair
+}
+
+// ServerContext implements the functionality of a Server.
+// TODO: add FullEvaluate
+type ServerContext interface {
+	// Evaluate evaluates the token
+	Evaluate(b BlindToken) (*Evaluation, error)
+}
+
+func generateCtx(suiteID uint16) string {
+	ctx := "000"
```

Comment:
nit: can we make this ("modeBase") a named constant?

Reply:
Sure!